projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php

```diff
@@ -474,18 +500,363 @@ private function unregister_disallowed_plugin_assets() {
+						return true;
+					}
+				}
+				continue; // Skip to next rule after checking core
+			}
+
+			// Check if handle starts with the rule (plugin handle prefix)
+			if ( str_starts_with( $handle, $rule . '-' ) ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Filters assets from HTML based on exclusion rules.
+	 *
+	 * @param string $html The HTML content to filter.
+	 * @param string $tag_name The HTML tag name to filter ('link' or 'script').
+	 * @param string $url_attribute The attribute containing the URL ('href' or 'src').
+	 * @param array  $exclude_rules Array of exclusion rules.
+	 * @return string The filtered HTML content.
+	 */
+	private function filter_assets_from_html( $html, $tag_name, $url_attribute, $exclude_rules ) {
+		if ( empty( $html ) || empty( $exclude_rules ) ) {
+			return $html;
+		}
+
+		// First, handle conditional comments separately (they're not parsed by DOMDocument)
+		$html = $this->filter_conditional_comments( $html, $tag_name, $url_attribute, $exclude_rules );
+
+		// Suppress warnings for malformed HTML
+		libxml_use_internal_errors( true );
+
+		$dom = new DOMDocument();
+		// Use UTF-8 encoding and load HTML fragment without adding doctype/html/body wrappers
+		$dom->loadHTML(
+			'<?xml encoding="UTF-8">' . $html,
+			LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD
+		);
+
+		// Remove the XML encoding processing instruction
+		// phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+		foreach ( $dom->childNodes as $node ) {
+			// phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+			if ( $node->nodeType === XML_PI_NODE ) {
+				$dom->removeChild( $node );
+				break;
+			}
+		}
+
+		// Process <link> tags (and <style> when filtering styles)
+		if ( 'link' === $tag_name ) {
+			$this->filter_link_elements( $dom, $url_attribute, $exclude_rules );
+			$this->filter_style_elements( $dom, $exclude_rules );
+		}
+
+		// Process <script> tags
+		if ( 'script' === $tag_name ) {
+			$this->filter_script_elements( $dom, $url_attribute, $exclude_rules );
+		}
+
+		libxml_clear_errors();
+
+		return $dom->saveHTML();
+	}
+
+	/**
+	 * Filters assets from conditional comments (<!--[if ...]>).
+	 *
+	 * IE conditional comments are not parsed as DOM elements by DOMDocument - they
+	 * remain as DOMComment nodes with HTML as plain text. This means we must use
+	 * regex to parse their content before DOM processing. This is the standard
+	 * approach for handling conditional comments across all HTML parsers.
+	 *
+	 * @param string $html The HTML content.
+	 * @param string $tag_name The HTML tag name ('link' or 'script').
+	 * @param string $url_attribute The attribute containing the URL ('href' or 'src').
+	 * @param array  $exclude_rules Array of exclusion rules.
+	 * @return string The filtered HTML content.
+	 */
+	private function filter_conditional_comments( $html, $tag_name, $url_attribute, $exclude_rules ) {
```

Comment:
This function's complexity and usage of regex is unfortunate, but necessary as `DOMDocument` cannot access the HTML tags within the conditional HTML comment.